Objects/obmalloc.c

```diff
@@ -103,6 +109,37 @@ _PyMem_RawFree(void *ctx, void *ptr)
     free(ptr);
 }
 
+static void *
+_PyMem_RawAlignedAlloc(void *ctx, size_t alignment, size_t size)
+{
+    if (size == 0) {
+        size = 1;
+    }
+
+#ifdef MS_WINDOWS
+    return _aligned_malloc(size, alignment);
+#else
+    void *ptr;
+
+    if (posix_memalign(&ptr, alignment, size)) {
+        return NULL;
+    }
+    /* posix_memalign() should not return NULL on success */
```

Comment:
Thanks for the change. Still, why would the function return NULL on success if size is guaranteed to be > 0?

posix_memalign() would be spectacularly broken if it did that.

So I find the comment and assert() confusing.

Reply:
I removed the comment but kept the assertion *just in case*.

I prefer an assertion to get an helpful error message rather than a crash in a random part of the code :-(

Reply:
There would be no crash.  People have to check the return value
of `_PyMem_RawAlignedAlloc()` for `NULL` in any case.

The thing is: Unusual `asserts()` have a cost for people reading the code. The reader wonders "does the Python core developer know something I don't".

Reply:
"There would be no crash. People have to check the return value of _PyMem_RawAlignedAlloc() for NULL in any case."

If posix_memalign() returns NULL on success, but your code raises a MemoryError exception if the pointer is NULL, you leak memory and your fails whereas the allocation succeeded.

This case should never occur. I prefer to be super careful and make sure that posix_memalign() never returns NULL on success.

> Unusual asserts() have a cost for people reading the code.

Do you want me to add a comment explaining the assertion?

Reply:
Yes, please add something like:

"This should never happen since we check for size > 0".

Also, pedantically, if the `assert()` is there then I'd use `void *ptr = NULL;` in the declaration.  Because if we check "can never happen" conditions we might as well check that `posix_memalign()` sets the pointer at all.

Reply:
Done.